### PR TITLE
[WIP] [SMTChecker] Refactor smt sorts and add Array

### PR DESCRIPTION
**Pull request by leonardoalt**
Depends on #5307 

This PR refactors the way SMT sorts are created/used in the SMTChecker.

`ArraySort` has two recursive subsorts: `domain` is the type of the indices and `range` is the type of the values.

`FunctionSort` has a vector of sorts `domain` for its parameters and a return sort `codomain`.

This PR is being split into smaller ones.

### CODE REVIEW

**Comment by leonardoalt** [comment]
Missing: add sort `Function`

**Comment by leonardoalt** [comment]
@ekpyron Will add the Function sort stuff in this PR instead of the UF one

File: libsolidity/formal/SolverInterface.h [hunk 1]
```diff
@@ -42,21 +42,65 @@ enum class CheckResult
...
+
+struct Sort
+{
+	Sort(Kind _kind):
+		kind(_kind) {}
+	virtual ~Sort() = default;
+	Kind const kind;
+	bool operator==(Sort const& _other) const { return kind == _other.kind; }
+};
+using SortPointer = std::shared_ptr<Sort>;
```
Comment (axic) [inline libsolidity/formal/SolverInterface.h:61]: I think you can make a refactoring PR which only introduces `struct Sort`.

File: libsolidity/formal/SolverInterface.h [hunk 2]
```diff
@@ -42,21 +42,65 @@ enum class CheckResult
...
 	friend class SolverInterface;
 public:
-	explicit Expression(bool _v): name(_v ? "true" : "false"), sort(Sort::Bool) {}
-	Expression(size_t _number): name(std::to_string(_number)), sort(Sort::Int) {}
-	Expression(u256 const& _number): name(_number.str()), sort(Sort::Int) {}
-	Expression(bigint const& _number): name(_number.str()), sort(Sort::Int) {}
+	explicit Expression(bool _v): Expression(_v ? "true" : "false", Kind::Bool) {}
+	Expression(size_t _number): Expression(std::to_string(_number), Kind::Int) {}
+	Expression(u256 const& _number): Expression(_number.str(), Kind::Int) {}
+	Expression(bigint const& _number): Expression(_number.str(), Kind::Int) {}
```
Comment (axic) [inline libsolidity/formal/SolverInterface.h:103]: Using this other constructor would belong in its own commit. Can be part of the "sort" cleanup PR mentioned above.

File: libsolidity/formal/SymbolicTypes.cpp [hunk 3]
```diff
@@ -54,6 +94,8 @@ pair<bool, shared_ptr<SymbolicVariable>> dev::solidity::newSymbolicVariable(
 		abstract = true;
 		var = make_shared<SymbolicIntVariable>(make_shared<IntegerType>(256), _uniqueName, _solver);
 	}
+	else if (isMapping(_type.category()))
```
Comment (axic) [inline libsolidity/formal/SymbolicTypes.cpp:97]: I don't think this is part of this PR?

File: libsolidity/formal/SolverInterface.h (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/formal/SolverInterface.h:61]: Wouldn't it be too small? That would change almost nothing. The main change is in how Expression uses Sort/ArraySort.

File: libsolidity/formal/SolverInterface.h (hunk 2 above)
Comment (leonardoalt) [reply libsolidity/formal/SolverInterface.h:103]: Makes sense

File: libsolidity/formal/SymbolicTypes.cpp (hunk 3 above)
Comment (leonardoalt) [reply libsolidity/formal/SymbolicTypes.cpp:97]: Yes, will remove

File: libsolidity/formal/SolverInterface.h (hunk 1 above)
Comment (axic) [reply libsolidity/formal/SolverInterface.h:61]: Small? All the interfaces changed from enum to struct :)

Comment (leonardoalt) [reply libsolidity/formal/SolverInterface.h:61]: Yes, if we include all those I agree.

Comment (leonardoalt) [reply libsolidity/formal/SolverInterface.h:61]: Created #5466

**Comment by axic** [comment]
Please rebase.

**Comment by axic** [comment]
After #5478 is anything left in this PR? Can this be closed?

**Comment by leonardoalt** [comment]
There are still some bits that need to be turned into PRs. Will close after.

**Comment by leonardoalt** [comment]
Closing in favor of the splits
